File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/event/AppEventClient.java
package com.adadapted.android.sdk.core.event;

import android.util.Log;

import com.adadapted.android.sdk.core.concurrency.ThreadPoolInteractorExecuter;
import com.adadapted.android.sdk.core.device.DeviceInfo;
import com.adadapted.android.sdk.core.device.DeviceInfoClient;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class AppEventClient {
    private static final String LOGTAG = AppEventClient.class.getName();

    private static class Types {
        static final String SDK = "sdk";
        static final String APP = "app";
    }

    private static AppEventClient instance;

    public static void createInstance(final AppEventSink sink) {
        if(instance == null) {
            instance = new AppEventClient(sink);
        }
    }

    private static AppEventClient getInstance() {
        return instance;
    }

    public static synchronized void trackAppEvent(final String name,
                                                  final Map<String, String> params) {
        if(instance == null) {
            return;
        }

        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable(){
            @Override
            public void run() {
                getInstance().performTrackEvent(Types.APP, name, params);
            }
        });
    }

    public static synchronized void trackAppEvent(final String name) {
        trackAppEvent(name, new HashMap<String, String>());
    }

    public static synchronized void trackSdkEvent(final String name,
                                                  final Map<String, String> params) {
        if(instance == null) {
            return;
        }

        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable(){
            @Override
            public void run() {
                getInstance().performTrackEvent(Types.SDK, name, params);
            }
        });
    }

    public static synchronized void trackSdkEvent(final String name) {
        trackSdkEvent(name, new HashMap<String, String>());
    }

    public static synchronized void trackError(final String code,
                                                  final String message,
                                                  final Map<String, String> params) {
        if(instance == null) {
            return;
        }

        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable(){
            @Override
            public void run() {
                getInstance().performTrackError(code, message, params);
            }
        });
    }

    public static synchronized void trackError(final String code,
                                               final String message) {
        trackError(code, message, new HashMap<String, String>());
    }

    public static synchronized void publishEvents() {
        if(instance == null) {
            return;
        }

        ThreadPoolInteractorExecuter.getInstance().executeInBackground(new Runnable() {
            @Override
            public void run() {
                getInstance().performPublishEvents();
                getInstance().performPublishErrors();
            }
        });
    }

    private final AppEventSink sink;

    private final Set<AppEvent> events;
    private final Lock eventLock = new ReentrantLock();

    private final Set<AppError> errors;
    private final Lock errorLock = new ReentrantLock();

    private AppEventClient(final AppEventSink sink) {
        this.sink = sink;

        this.events = new HashSet<>();
        this.errors = new HashSet<>();

        DeviceInfoClient.getDeviceInfo(new DeviceInfoClient.Callback() {
            @Override
            public void onDeviceInfoCollected(final DeviceInfo deviceInfo) {
                sink.generateWrappers(deviceInfo);
            }
        });
    }

    private void performTrackEvent(final String type,
                                   final String name,
                                   final Map<String, String> params) {
        //Log.i(LOGTAG, "App Event: " + type + " - " + name);
        eventLock.lock();
        try {
            events.add(new AppEvent(type, name, params));
        }
        finally {
            eventLock.unlock();
        }
    }

    private void performTrackError(final String code,
                                   final String message,
                                   final Map<String, String> params) {

        Log.w(LOGTAG, "App Error: " + code + " - " + message);
        errorLock.lock();
        try {
            errors.add(new AppError(code, message, params));
        }
        finally {
            errorLock.unlock();
        }
    }

    private void performPublishEvents() {
        eventLock.lock();
        try {
            if(!events.isEmpty()) {
                final Set<AppEvent> currentEvents = new HashSet<>(events);
                events.clear();
                sink.publishEvent(currentEvents);
            }
        } finally{
            eventLock.unlock();
        }
    }

    private void performPublishErrors() {
        errorLock.lock();
        try {
            if(!errors.isEmpty()) {
                final Set<AppError> currentErrors = new HashSet<>(errors);
                errors.clear();

                sink.publishError(currentErrors);
            }
        }
        finally {
            errorLock.unlock();
        }
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/event/AppEventClient.kt
package com.adadapted.android.sdk.core.event

import android.util.Log
import com.adadapted.android.sdk.config.EventStrings
import com.adadapted.android.sdk.core.concurrency.TransporterCoroutineScope
import com.adadapted.android.sdk.core.device.DeviceInfo
import com.adadapted.android.sdk.core.device.DeviceInfoClient
import com.adadapted.android.sdk.core.session.SessionClient
import com.adadapted.android.sdk.core.session.SessionListener
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

class AppEventClient private constructor(private val sink: AppEventSink, private val transporter: TransporterCoroutineScope): SessionListener() {
    private object Types {
        const val SDK = "sdk"
        const val APP = "app"
    }

    private val events: MutableSet<AppEvent>
    private val eventLock: Lock = ReentrantLock()
    private val errors: MutableSet<AppError>
    private val errorLock: Lock = ReentrantLock()
    private fun performTrackEvent(type: String, name: String, params: Map<String, String>) {
        eventLock.lock()
        try {
            events.add(AppEvent(type, name, params))
        } finally {
            eventLock.unlock()
        }
    }

    private fun performTrackError(code: String, message: String, params: Map<String, String>) {
        Log.w(LOGTAG, "App Error: $code - $message")
        errorLock.lock()
        try {
            errors.add(AppError(code, message, params))
        } finally {
            errorLock.unlock()
        }
    }

    private fun performPublishEvents() {
        eventLock.lock()
        try {
            if (events.isNotEmpty()) {
                val currentEvents: Set<AppEvent> = HashSet(events)
                events.clear()
                sink.publishEvent(currentEvents)
            }
        } finally {
            eventLock.unlock()
        }
    }

    private fun performPublishErrors() {
        errorLock.lock()
        try {
            if (errors.isNotEmpty()) {
                val currentErrors: Set<AppError> = HashSet(errors)
                errors.clear()
                sink.publishError(currentErrors)
            }
        } finally {
            errorLock.unlock()
        }
    }

    override fun onPublishEvents() {
        transporter.dispatchToBackground {
            performPublishEvents()
            performPublishErrors()
        }
    }

    override fun onSessionExpired() {
        trackSdkEvent(EventStrings.EXPIRED_EVENT)
    }

    @Synchronized
    fun trackSdkEvent(name: String, params: Map<String, String> = HashMap()) {
        transporter.dispatchToBackground { performTrackEvent(Types.SDK, name, params) }
    }

    @Synchronized
    fun trackError(code: String, message: String, params: Map<String, String> = HashMap()) {
        transporter.dispatchToBackground {
            performTrackError(code, message, params)
        }
    }

    @Synchronized
    fun trackAppEvent(name: String, params: Map<String, String> = HashMap()) {
        transporter.dispatchToBackground {
            performTrackEvent(Types.APP, name, params)
        }
    }

    companion object {
        private val LOGTAG = AppEventClient::class.java.name
        private lateinit var instance: AppEventClient

        fun createInstance(sink: AppEventSink, transporter: TransporterCoroutineScope) {
            instance = AppEventClient(sink, transporter)
        }

        fun getInstance(): AppEventClient {
            return instance
        }
    }

    init {
        events = HashSet()
        errors = HashSet()

        DeviceInfoClient.getInstance().getDeviceInfo(object: DeviceInfoClient.Callback {
            override fun onDeviceInfoCollected(deviceInfo: DeviceInfo) {
                sink.generateWrappers(deviceInfo)
            }
        })

        SessionClient.getInstance().addListener(this)
    }
}